var/spack/repos/builtin/packages/hdf5/package.py

```diff
@@ -257,85 +268,86 @@ def libs(self):
+        args = [
+            self.define('ALLOW_UNSUPPORTED', True),
+            self.define('BUILD_TESTING', self.run_tests),
+            self.define('HDF5_ENABLE_Z_LIB_SUPPORT', True),
+            self.define_from_variant('HDF5_ENABLE_SZIP_SUPPORT', 'szip'),
+            self.define_from_variant('BUILD_SHARED_LIBS', 'shared'),
+            self.define_from_variant('ONLY_SHARED_LIBS', 'shared'),
+            self.define_from_variant('HDF5_ENABLE_PARALLEL', 'mpi'),
+            self.define_from_variant('HDF5_ENABLE_THREADSAFE', 'threadsafe'),
+            self.define_from_variant('HDF5_BUILD_HL_LIB', 'hl'),
+            self.define_from_variant('HDF5_BUILD_CPP_LIB', 'cxx'),
+            self.define_from_variant('HDF5_BUILD_FORTRAN', 'fortran'),
+            self.define_from_variant('HDF5_BUILD_JAVA', 'java'),
+            self.define_from_variant('HDF5_BUILD_TOOLS', 'tools')
+        ]
 
-        if '+shared' in self.spec:
-            extra_args.append('--enable-shared')
-        else:
-            extra_args.append('--disable-shared')
-            extra_args.append('--enable-static-exec')
+        if '+szip' in spec:
+            args.append(self.define('HDF5_ENABLE_SZIP_ENCODING', True))
+
+        api = spec.variants['api'].value
+        if api != 'default':
+            args.append(self.define('DEFAULT_API_VERSION', api))
+
+        # The variable CMAKE_POSITION_INDEPENDENT_CODE is set to True by default in
+        # HDF5 Cmake code;  this should insure that it is off if '~pic' in spec. The
+        # flags that were previously set for '+pic' in the Autotools version are kept
+        # in case CMake doesn't set them as expected.  Both the 'pic' variant and the
+        # CMAKE_POSITION_INDEPENDENT_CODE variable are True by default, but '~pic'
+        # should disable them both.
+        args.append(self.define_from_variant('CMAKE_POSITION_INDEPENDENT_CODE', 'pic'))
+        if '+pic' in spec:
+            # use global spack compiler flags
+            _flags = self.compiler.cc_pic_flag
+            _flags += " " + ' '.join(spec.compiler_flags['cflags'])
+            args.append('CFLAGS={0}'.format(_flags))
+
+            if '+cxx' in spec:
+                _flags = self.compiler.cxx_pic_flag
+                _flags += " " + ' '.join(spec.compiler_flags['cxxflags'])
+                args.append('CXXFLAGS={0}'.format(_flags))
+
+            if '+fortran' in spec:
+                _flags = self.compiler.fc_pic_flag
+                _flags += " " + ' '.join(spec.compiler_flags['fflags'])
+                args.append('FCFLAGS={0}'.format(_flags))
 
         # Fujitsu Compiler does not add Fortran runtime in rpath.
-        if '+fortran %fj' in self.spec:
-            extra_args.append('LDFLAGS=-lfj90i -lfj90f -lfjsrcinfo -lelf')
+        if '+fortran %fj' in spec:
+            args.append('LDFLAGS=-lfj90i -lfj90f -lfjsrcinfo -lelf')
 
-        if '+mpi' in self.spec:
+        if '+mpi' in spec:
```

Comment:
The same problem here: `CC`, `CXX` and `FC` are ignored by Cmake when set on the command line. The code that runs under this condition could either be removed or replaced with the code in my comment regarding the patches.